diff --git a/mlir/lib/Dialect/AffineOps/AffineOps.cpp b/mlir/lib/Dialect/AffineOps/AffineOps.cpp
index 5f4cc2e1060..916e8881210 100644
--- a/mlir/lib/Dialect/AffineOps/AffineOps.cpp
+++ b/mlir/lib/Dialect/AffineOps/AffineOps.cpp
@@ -134,8 +134,9 @@ bool mlir::isValidDim(Value value) {
     return false;
   }
   // This value has to be a block argument for a FuncOp or an affine.for.
-  auto *parentOp = value.cast<BlockArgument>()->getOwner()->getParentOp();
-  return isa<FuncOp>(parentOp) || isa<AffineForOp>(parentOp);
+  // auto *parentOp = value.cast<BlockArgument>()->getOwner()->getParentOp();
+  // return isa<FuncOp>(parentOp) || isa<AffineForOp>(parentOp);
+  return true;
 }

 /// Returns true if the 'index' dimension of the `memref` defined by
 